### PR TITLE
CORDA-716 Make serialization init less static

### PR DESCRIPTION
**Pull request by andr3ej**
docs to be added after initial reviews

### CODE REVIEW

**Comment by andr3ej** [comment]
@shamsasari this may interest you in relation to smoketests

File: webserver/src/integration-test/kotlin/net/corda/webserver/WebserverDriverTests.kt [hunk 1]
```diff
@@ -11,7 +11,7 @@ import org.junit.Test
 import java.util.concurrent.Executors
 import java.util.concurrent.ScheduledExecutorService
 
-class DriverTests {
+class WebserverDriverTests {
```
Comment (andr3ej) [inline webserver/src/integration-test/kotlin/net/corda/webserver/WebserverDriverTests.kt]: unrelated naming fix

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/KryoClientSerializationScheme.kt [hunk 2]
```diff
@@ -29,25 +31,15 @@ class KryoClientSerializationScheme : AbstractKryoSerializationScheme() {
...
-                SerializationDefaults.RPC_CLIENT_CONTEXT = KRYO_RPC_CLIENT_CONTEXT
-            } catch (e: IllegalStateException) {
-                // Check that it's registered as we expect
-                val factory = SerializationDefaults.SERIALIZATION_FACTORY
-                val checkedFactory = factory as? SerializationFactoryImpl
-                        ?: throw IllegalStateException("RPC client encountered conflicting configuration of serialization subsystem: $factory")
-                check(checkedFactory.alreadyRegisteredSchemes.any { it is KryoClientSerializationScheme }) {
-                    "RPC client encountered conflicting configuration of serialization subsystem."
-                }
-            }
```
Comment (andr3ej) [inline client/rpc/src/main/kotlin/net/corda/client/rpc/internal/KryoClientSerializationScheme.kt]: this logic replaced by ToggleField and stricter SerializationDefaults

File: client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClient.kt [hunk 3]
```diff
@@ -110,6 +111,7 @@ class RPCClient<I : RPCOps>(
                 maxRetryInterval = rpcConfiguration.connectionMaxRetryInterval.toMillis()
                 reconnectAttempts = rpcConfiguration.maxReconnectAttempts
                 minLargeMessageSize = rpcConfiguration.maxFileSize
+                isUseGlobalPools = nodeSerializationEnv != null
```
Comment (andr3ej) [inline client/rpc/src/main/kotlin/net/corda/client/rpc/internal/RPCClient.kt]: static pools interfere with inheritable serialization env, so only enable them when not testing

File: core/src/main/kotlin/net/corda/core/internal/ToggleField.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    override fun get() = threadLocal.get()?.valueOrDeclareLeak()
+    override fun setImpl(value: T) = threadLocal.set(Holder(value))
+    override fun clear() = threadLocal.run {
+        val holder = get()!!
+        remove()
+        holder.set(null) // Threads that inherited the holder are now considered to have escaped from the test.
+    }
+}
```
Comment (andr3ej) [inline core/src/main/kotlin/net/corda/core/internal/ToggleField.kt]: the unit tests explain what's going on here

File: core/src/main/kotlin/net/corda/core/serialization/SerializationAPI.kt [hunk 5]
```diff
@@ -170,17 +173,48 @@ interface SerializationContext {
...
+    override lateinit var STORAGE_CONTEXT: SerializationContext
+    override lateinit var CHECKPOINT_CONTEXT: SerializationContext
+
+    init {
+        rpcServerContext?.let { RPC_SERVER_CONTEXT = it }
+        rpcClientContext?.let { RPC_CLIENT_CONTEXT = it }
+        storageContext?.let { STORAGE_CONTEXT = it }
+        checkpointContext?.let { CHECKPOINT_CONTEXT = it }
+    }
+}
```
Comment (andr3ej) [inline core/src/main/kotlin/net/corda/core/serialization/SerializationAPI.kt]: this breaks api, and i don't know what the right answer is, suggestions welcome

File: core/src/main/kotlin/net/corda/core/serialization/SerializationAPI.kt [hunk 6]
```diff
@@ -170,17 +173,48 @@ interface SerializationContext {
...
+
+private val _nodeSerializationEnv = SimpleToggleField<SerializationEnvironment>("nodeSerializationEnv")
+@VisibleForTesting
+val _globalSerializationEnv = SimpleToggleField<SerializationEnvironment>("globalSerializationEnv")
```
Comment (andr3ej) [inline core/src/main/kotlin/net/corda/core/serialization/SerializationAPI.kt]: this might need a better name, it's like nodeSerializationEnv but not permanent

File: node/src/test/kotlin/net/corda/node/services/vault/VaultWithCashTest.kt [hunk 7]
```diff
@@ -150,82 +154,74 @@ class VaultWithCashTest {
...
         }
-
-        countDown.await()
+        val both = listOf(first, second).transpose()
+        try {
+            both.getOrThrow()
+            fail("Expected insufficient balance.")
+        } catch (e: InsufficientBalanceException) {
+            assertEquals(0, e.suppressed.size) // One should succeed.
+        }
```
Comment (andr3ej) [inline node/src/test/kotlin/net/corda/node/services/vault/VaultWithCashTest.kt]: it was hard to see what was going wrong in this test so i fixed it

File: testing/test-utils/src/main/kotlin/net/corda/testing/SerializationTestHelpers.kt [hunk 8]
```diff
@@ -15,183 +13,84 @@ import org.junit.rules.TestRule
...
+    } finally {
+        property.set(null)
+    }
+}
+
+/**
+ * For example your test class uses [SerializationEnvironmentRule] but you want to turn it off for one method.
+ * Use sparingly, ideally a test class shouldn't mix serialization init mechanisms.
+ */
+fun <T> withoutTestSerialization(callable: () -> T): T {
```
Comment (andr3ej) [inline testing/test-utils/src/main/kotlin/net/corda/testing/SerializationTestHelpers.kt]: only used in one place, i might get rid of it in a subsequent pr

**Comment by andr3ej** [comment]
todo: java example of serialize(context = SerializationDefaults.whatever) for api

File: core/src/main/kotlin/net/corda/core/serialization/internal/SerializationEnvironment.kt [hunk 9]
```diff
@@ -1,13 +1,57 @@
...
+        storageContext: SerializationContext? = null,
+        checkpointContext: SerializationContext? = null) : SerializationEnvironment {
+    // Those that are passed in as null are never inited:
+    override lateinit var rpcServerContext: SerializationContext
+    override lateinit var rpcClientContext: SerializationContext
+    override lateinit var storageContext: SerializationContext
+    override lateinit var checkpointContext: SerializationContext
+
+    init {
+        rpcServerContext?.let { this.rpcServerContext = it }
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/serialization/internal/SerializationEnvironment.kt]: Nice.  So this sort of logically ORs the contexts, so client + server works.

Comment (rick-r3) [reply core/src/main/kotlin/net/corda/core/serialization/internal/SerializationEnvironment.kt]: But what about the factory?  Is that where the thread local stuff comes in ?

Comment (andr3ej) [reply core/src/main/kotlin/net/corda/core/serialization/internal/SerializationEnvironment.kt]: i'm not sure what you mean. the environment is the global serialization config. the thread local stuff is for tests so that each test has its own idea of what the global serialization config is.

File: client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt [hunk 10]
```diff
@@ -70,11 +68,6 @@ class CordaRPCClient @JvmOverloads constructor(
         hostAndPort: NetworkHostAndPort,
         configuration: CordaRPCClientConfiguration = CordaRPCClientConfiguration.DEFAULT
 ) {
-    init {
```
Comment (rick-r3) [inline client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt]: As per `WebServer`, do clients now need to initialize the serialization themselves if using RPC client?  That seems like a sharp edge.  Can we not detect if there's something initialised, and if not, do what it used to do?

File: core/src/main/kotlin/net/corda/core/serialization/internal/SerializationEnvironment.kt (hunk 9 above)
Comment (rick-r3) [reply core/src/main/kotlin/net/corda/core/serialization/internal/SerializationEnvironment.kt]: Yep.  I found the bit of code where it selects the thread local contexts etc.  Ignore my previous comment, above.

File: client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt (hunk 10 above)
Comment (andr3ej) [reply client/rpc/src/main/kotlin/net/corda/client/rpc/CordaRPCClient.kt]: done. in a follow-up pr i may shift the multiple-environment check to be on-write, which will make this bit of code nicer and the read case more performant.
